pkg/proxy/winuserspace/proxier.go

```diff
@@ -304,322 +322,85 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
+			// create a slice of all the source IPs to use for service port portals
+			listenIPPortMap := getListenIPPortMap(service, int(servicePort.Port), int(servicePort.NodePort))
+			protocol := servicePort.Protocol
+
+			for listenIP, listenPort := range listenIPPortMap {
+				servicePortPortalName := ServicePortPortalName{
+					NamespacedName: types.NamespacedName{
+						Namespace: service.Namespace,
+						Name:      service.Name},
+					Port:         servicePort.Name,
+					PortalIPName: listenIP,
+				}
+				activeServicePortPortals[servicePortPortalName] = true
+				info, exists := proxier.getServiceInfo(servicePortPortalName)
+				if exists && sameConfig(info, service, protocol, listenPort) {
+					// Nothing changed.
+					continue
+				}
+				if exists {
+					glog.V(4).Infof("Something changed for service %q: stopping it", servicePortPortalName)
+					if err := proxier.closeServicePortPortal(servicePortPortalName, info); err != nil {
+						glog.Errorf("Failed to close service port portal %q: %v", servicePortPortalName, err)
+					}
 				}
-				err = proxier.stopProxy(serviceName, info)
+				glog.V(1).Infof("Adding new service %q at %s:%d/%s", servicePortPortalName, listenIP, listenPort, protocol)
+				info, err := proxier.addServicePortPortal(servicePortPortalName, protocol, listenIP, listenPort, proxier.udpIdleTimeout)
 				if err != nil {
-					glog.Errorf("Failed to stop service %q: %v", serviceName, err)
+					glog.Errorf("Failed to start proxy for %q: %v", servicePortPortalName, err)
+					continue
 				}
+				info.sessionAffinityType = service.Spec.SessionAffinity
+				glog.V(10).Infof("info: %#v", info)
 			}
-
-			proxyPort, err := proxier.proxyPorts.AllocateNext()
-			if err != nil {
-				glog.Errorf("failed to allocate proxy port for service %q: %v", serviceName, err)
-				continue
-			}
-
-			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, serviceIP, servicePort.Port, servicePort.Protocol)
-			info, err = proxier.addServiceOnPort(serviceName, servicePort.Protocol, proxyPort, proxier.udpIdleTimeout)
-			if err != nil {
-				glog.Errorf("Failed to start proxy for %q: %v", serviceName, err)
-				continue
-			}
-			info.portal.ip = serviceIP
-			info.portal.port = int(servicePort.Port)
-			info.externalIPs = service.Spec.ExternalIPs
-			// Deep-copy in case the service instance changes
-			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
-			info.nodePort = int(servicePort.NodePort)
-			info.sessionAffinityType = service.Spec.SessionAffinity
-			glog.V(4).Infof("info: %#v", info)
-
-			err = proxier.openPortal(serviceName, info)
-			if err != nil {
-				glog.Errorf("Failed to open portal for %q: %v", serviceName, err)
+			if len(listenIPPortMap) > 0 {
+				// only one loadbalancer per service port portal
+				servicePortName := proxy.ServicePortName{
+					NamespacedName: types.NamespacedName{
+						Namespace: service.Namespace,
+						Name:      service.Name},
```

Comment:
I did a search through all my changes and fixed '},' that didn't have line break